### PR TITLE
Improve bound variable checking in CREATE clauses

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Improve CREATE validations, fixes crash in #817

### CODE REVIEW

File: tests/flow/test_pagerank.py [hunk 1]
```diff
@@ -17,30 +17,30 @@ def __init__(self):
...
         queries = [
             # Single Label, single connection.
-            "CREATE (a:L {v:1})-[:R]->(b:L {v:2})"
+            "CREATE (a:L {v:1})-[:R]->(b:L {v:2})",
             # Single Label, multi connection.
-            "CREATE (a:L {v:1})-[:R]->(b:L {v:2}), (a)-[:R]->(b)"
+            "CREATE (a:L {v:1})-[:R]->(b:L {v:2}), (a)-[:R]->(b)",
             # Multi Label, single connection.
-            "CREATE (a:L {v:1})-[:R]->(b:L {v:2}), (:X)-[:R]->(:X)"
+            "CREATE (a:L {v:1})-[:R]->(b:L {v:2}), (:X)-[:R]->(:X)",
```
Comment (jeffreylovitz) [inline tests/flow/test_pagerank.py]: The changes in this file are not required by this fix, but previously all of these strings were implicitly concatenated into a single query.

File: src/ast/ast_validations.c [hunk 2]
```diff
@@ -762,8 +762,8 @@ static AST_Validation _Validate_MERGE_Clauses(const AST *ast, char **reason) {
 		uint clause_idx = merge_clause_indices[i];
 
 		// Collect all entities that are bound before this MERGE clause.
-		for(uint j = start_offset; i < clause_idx; i ++) {
-			const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
+		for(uint j = start_offset; j < clause_idx; j ++) {
```
Comment (swilly22) [inline src/ast/ast_validations.c]: Oh no!

File: src/ast/ast_validations.c [hunk 3]
```diff
@@ -790,36 +790,39 @@ static AST_Validation _Validate_MERGE_Clauses(const AST *ast, char **reason) {
...
+	if(_ValidateFunctionCalls(pattern, reason, false) != AST_VALID) return AST_INVALID;
+
 	uint path_count = cypher_ast_pattern_npaths(pattern);
 	for(uint i = 0; i < path_count; i ++) {
 		const cypher_astnode_t *path = cypher_ast_pattern_get_path(pattern, i);
+		// Validate that inlined properties are valid.
 		if(_ValidateInlinedPropertiesOnPath(path, reason) != AST_VALID) return AST_INVALID;
+
+		uint nelems = cypher_ast_pattern_path_nelements(path);
+		for(uint j = 1; j < nelems; j += 2) {
```
Comment (swilly22) [inline src/ast/ast_validations.c]: Can you please explain (add comment) this loop setup?

File: src/ast/ast_validations.c [hunk 4]
```diff
@@ -830,21 +833,23 @@ static AST_Validation _Validate_CREATE_Clauses(const AST *ast, char **reason) {
...
+		for(uint j = start_offset; j < clause_idx; j ++) {
+			const cypher_astnode_t *prev_clause = cypher_ast_query_get_clause(ast->root, i);
+			_AST_GetDefinedIdentifiers(prev_clause, defined_aliases);
 		}
+		start_offset = clause_idx;
 
-		// Validate that inlined properties do not use undefined identifiers.
-		res = _Validate_CREATE_Clause_Properties(clause, reason);
+		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, clause_idx);
+		res = _Validate_CREATE_Entities(clause, defined_aliases, reason);
```
Comment (swilly22) [inline src/ast/ast_validations.c]: does `defined_aliases` gets cleared between each loop iteration?

File: tests/flow/test_pagerank.py (hunk 1 above)
Comment (swilly22) [reply tests/flow/test_pagerank.py]: good catch

File: src/ast/ast_validations.c (hunk 4 above)
Comment (jeffreylovitz) [reply src/ast/ast_validations.c]: No, each iteration adds all the aliases introduced from clauses[i-1] to clauses[i] to the same rax. We use similar logic in MERGE validation. This validation is used on scoped ASTs, so this is preferable to resetting.

File: src/ast/ast_validations.c (hunk 2 above)
Comment (jeffreylovitz) [reply src/ast/ast_validations.c]: I know! I can't believe this didn't trip us up before.

**Pull request merged**